File: src/naive_search.py
import time
import gzip
from pathlib import Path
import iv2py as iv

def load_fasta(filepath):
    sequences = []
    current_seq = []
    
    open_func = gzip.open if filepath.endswith('.gz') else open
    
    with open_func(filepath, 'rt') as f:
        line_count = 0
        for line in f:
            line = line.strip()
            line_count += 1
            
            if line.startswith('>'):
                if current_seq:
                    sequences.append(''.join(current_seq))
                    current_seq = []
            else:
                current_seq.append(line)
        
        if current_seq:
            sequences.append(''.join(current_seq))
    return sequences


def naive_search(query, reference):
    positions = []
    ref_len = len(reference)
    query_len = len(query)
    
    for i in range(ref_len - query_len + 1):
        if reference[i:i + query_len] == query:
            positions.append(i)
            
    return positions


def benchmark_algorithm(queries, reference):
    print(f"Running Naive Search")
    print(f"Processing {len(queries)} queries")
    
    start = time.time()
    total_hits = 0
    
    progress_interval = max(100, len(queries) // 100)
    
    for idx, query in enumerate(queries):
        hits = naive_search(query, reference)
        total_hits += len(hits)
        
        if (idx + 1) % progress_interval == 0 or idx == 0:
            elapsed_so_far = time.time() - start
            percent = ((idx + 1) / len(queries)) * 100
            queries_per_sec = (idx + 1) / elapsed_so_far if elapsed_so_far > 0 else 0
            print(f"Progress: {idx + 1:>10}/{len(queries)} ({percent:>5.1f}%) - {elapsed_so_far:>8.2f}s - {queries_per_sec:>8.1f} q/s - {total_hits} hits")
    
    elapsed = time.time() - start
    
    print(f"{algorithm_name} complete")
    print(f"Queries processed: {len(queries)}")
    print(f"Total hits found: {total_hits}")
    print(f"Total time: {elapsed:.4f}s")
    print(f"Time per query: {elapsed/len(queries)*1000:.4f}ms")
    print(f"Queries per second: {len(queries)/elapsed:.2f}")
    
    return elapsed, total_hits

# Load reference genome
reference_file = "hg38_partial.fasta.gz"
references = load_fasta(reference_file)

reference = references[0]
print(f"Reference genome: {len(reference):,} base pairs")

# Load queries of length 100 for benchmarking different query counts
queries_100 = load_fasta("illumina_reads_100.fasta.gz")
print(f"Total queries available (length 100): {len(queries_100):,}")

print(f"Testing with 100 queries (length 100)")
test_queries = queries_100[:100]
print("Naive Search")
benchmark_algorithm(test_queries, reference)

File: src/naive_search.cpp
#include <sstream>
#include <chrono>
#include <iomanip>

#include <seqan3/alphabet/nucleotide/dna5.hpp>
#include <seqan3/argument_parser/all.hpp>
#include <seqan3/core/debug_stream.hpp>
#include <seqan3/io/sequence_file/all.hpp>
#include <seqan3/search/fm_index/fm_index.hpp>
#include <seqan3/search/search.hpp>

// prints out all occurences of query inside of ref
size_t findOccurences(std::vector<seqan3::dna5> const& ref, std::vector<seqan3::dna5> const& query) {
    if (query.empty() || ref.empty() || query.size() > ref.size()) {
        return 0;
    }
    
    size_t match_count = 0;
    
    // Naive search: slide the query over the reference
    for (size_t i = 0; i <= ref.size() - query.size(); ++i) {
        bool match = true;
        
        // Check if query matches at position i
        for (size_t j = 0; j < query.size(); ++j) {
            if (ref[i + j] != query[j]) {
                match = false;
                break;
            }
        }
        
        if (match) {
            match_count++;
        }
    }
    
    return match_count;
}

int main(int argc, char const* const* argv) {
    seqan3::argument_parser parser{"naive_search", argc, argv, seqan3::update_notifications::off};

    parser.info.author = "SeqAn-Team";
    parser.info.version = "1.0.0";

    auto reference_file = std::filesystem::path{};
    parser.add_option(reference_file, '\0', "reference", "path to the reference file");

    auto query_file = std::filesystem::path{};
    parser.add_option(query_file, '\0', "query", "path to the query file");

    auto number_of_queries = size_t{100};
    parser.add_option(number_of_queries, '\0', "query_ct", "number of query, if not enough queries, these will be duplicated");

    try {
         parser.parse();
    } catch (seqan3::argument_parser_error const& ext) {
        seqan3::debug_stream << "Parsing error. " << ext.what() << "\n";
        return EXIT_FAILURE;
    }

    // loading our files
    auto reference_stream = seqan3::sequence_file_input{reference_file};
    auto query_stream     = seqan3::sequence_file_input{query_file};

    // read reference into memory
    std::vector<std::vector<seqan3::dna5>> reference;
    size_t total_ref_size = 0;
    for (auto& record : reference_stream) {
        reference.push_back(record.sequence());
        total_ref_size += record.sequence().size();
    }

    // read query into memory
    std::vector<std::vector<seqan3::dna5>> queries;
    for (auto& record : query_stream) {
        queries.push_back(record.sequence());
    }

    size_t query_length = queries.empty() ? 0 : queries[0].size();
    size_t available_queries = queries.size();

    // Use the requested number of queries (or available if less)
    number_of_queries = std::min(number_of_queries, available_queries);
    
    // Duplicate queries if needed
    while (queries.size() < number_of_queries) {
        auto old_count = queries.size();
        queries.resize(2 * old_count);
        std::copy_n(queries.begin(), old_count, queries.begin() + old_count);
    }
    queries.resize(number_of_queries);

    std::cout << "Reference genome: " << total_ref_size << " base pairs\n";
    std::cout << "Total queries available (length " << query_length << "): " << available_queries << "\n";
    std::cout << "Testing with " << number_of_queries << " queries (length " << query_length << ")\n";
    std::cout << "\n[Naive Search]\n";
    std::cout << "Running Naive Search\n";
    std::cout << "Processing " << number_of_queries << " queries\n";

    // Start timing - search
    auto start_search = std::chrono::high_resolution_clock::now();

    //! search for all occurences of queries inside of reference
    size_t total_matches = 0;
    size_t processed = 0;
    
    for (auto& r : reference) {
        for (auto& q : queries) {
            total_matches += findOccurences(r, q);
            processed++;
            
            // Print progress at first query and last query
            if (processed == 1 || processed == number_of_queries) {
                auto current_time = std::chrono::high_resolution_clock::now();
                auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(current_time - start_search).count() / 1000.0;
                double qps = processed / elapsed;
                
                std::cout << "Progress: " << std::setw(10) << processed << "/" << number_of_queries 
                          << " (" << std::fixed << std::setprecision(1) << (100.0 * processed / number_of_queries) << "%) - "
                          << std::setw(10) << std::fixed << std::setprecision(2) << elapsed << "s - "
                          << std::setw(10) << std::fixed << std::setprecision(1) << qps << " q/s - "
                          << total_matches << " hits\n";
            }
        }
    }

    auto end_search = std::chrono::high_resolution_clock::now();
    auto search_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_search - start_search).count() / 1000.0;

    std::cout << "\nSearch completed in " << std::fixed << std::setprecision(2) << search_time << "s\n";
    std::cout << "Total hits found: " << total_matches << "\n";

    return 0;
}